clang/lib/Sema/SemaSYCL.cpp

```diff
@@ -410,6 +474,36 @@ class OutlinedFunctionDeclBodyInstantiator
   ParmDeclMap &MapRef;
 };
 
+OutlinedFunctionDecl *BuildSYCLKernelEntryPointOutline(Sema &SemaRef,
+                                                       FunctionDecl *FD,
+                                                       CompoundStmt *Body) {
+  using ParmDeclMap = OutlinedFunctionDeclBodyInstantiator::ParmDeclMap;
+  ParmDeclMap ParmMap;
+
+  OutlinedFunctionDecl *OFD = OutlinedFunctionDecl::Create(
+      SemaRef.getASTContext(), FD, FD->getNumParams());
+  unsigned i = 0;
+  for (ParmVarDecl *PVD : FD->parameters()) {
+    ImplicitParamDecl *IPD = ImplicitParamDecl::Create(
+        SemaRef.getASTContext(), OFD, SourceLocation(), PVD->getIdentifier(),
+        PVD->getType(), ImplicitParamKind::Other);
+    OFD->setParam(i, IPD);
+    ParmMap[PVD] = IPD;
+    ++i;
+  }
+
+  // FIXME: Diagnose (implicit or explicit) use of CXXThisExpr in potentially
+  // evaluated contexts in the function body. This is not necessarily the
+  // right place to add such a diagnostic.
```

Comment:
@Fznamznon, I'll take care of diagnosing use of `this` if you don't beat me to it.

Reply:
Diagnostics for use of `this` have been completed; resolving this comment.